xtask/src/ci.rs

```diff
@@ -0,0 +1,86 @@
+use std::{env, path::PathBuf};
+
+use clap::{Args, Subcommand};
+use serde::Deserialize;
+use xshell::{cmd, pushd};
+
+use crate::{build_docs, Result};
+
+#[derive(Args)]
+pub struct CiArgs {
+    #[clap(subcommand)]
+    cmd: Option<CiCommand>,
+}
+
+#[derive(Subcommand)]
+enum CiCommand {
+    /// Check style
+    Style,
+    /// Check for typos
+    Typos,
+    /// Check clippy lints
+    Clippy,
+    /// Check documentation
+    Docs,
+}
+
+impl CiArgs {
+    pub fn run(self) -> Result<()> {
+        let _p = pushd(&workspace_root()?)?;
+
+        match self.cmd {
+            Some(cmd) => match cmd {
+                CiCommand::Style => check_style(),
+                CiCommand::Typos => check_typos(),
+                CiCommand::Clippy => check_clippy(),
+                CiCommand::Docs => check_docs(),
+            },
+            None => {
+                check_style()?;
+                check_clippy()?;
+                check_typos()?;
+                check_docs()?;
+
+                Ok(())
+            }
+        }
+    }
+}
+
+fn check_style() -> Result<()> {
+    cmd!("rustup run nightly cargo fmt -- --check").run()?;
+    Ok(())
+}
+
+fn check_typos() -> Result<()> {
+    // FIXME: Print install instructions if command-not-found (needs an xshell
+    //        change: https://github.com/matklad/xshell/issues/46)
+    cmd!("typos").run()?;
+    Ok(())
+}
+
+fn check_clippy() -> Result<()> {
+    cmd!("rustup run nightly cargo clippy --all-targets -- -D warnings").run()?;
+    cmd!(
+        "rustup run nightly cargo clippy --all-targets
+            --no-default-features --features native-tls,warp
+            -- -D warnings"
+    )
+    .run()?;
+    Ok(())
+}
+
+fn check_docs() -> Result<()> {
+    build_docs(["--", "-Dwarnings"])
```

Comment:
Does this one work for you? It throws an unhelpful error for me:

```
$ rustup run nightly cargo doc --no-deps --workspace --features docsrs -Zrustdoc-map -- -Dwarnings
error: Found argument '-Dwarnings' which wasn't expected, or isn't valid in this context

	If you tried to supply `-Dwarnings` as a value rather than a flag, use `-- -Dwarnings`
```